akka-actor/src/main/scala/akka/util/FrequencyList.scala

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright (C) 2021 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.util
+
+import akka.annotation.InternalApi
+
+import scala.collection.{ immutable, mutable, AbstractIterator }
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] object FrequencyList {
+  def empty[A]: FrequencyList[A] = new FrequencyList[A]
+
+  private final class FrequencyNode[A](val count: Int) {
+    var lessFrequent, moreFrequent: OptionVal[FrequencyNode[A]] = OptionVal.None
+    var leastRecent, mostRecent: OptionVal[Node[A]] = OptionVal.None
+  }
+
+  private final class Node[A](val value: A) {
+    var frequencyNode: OptionVal[FrequencyNode[A]] = OptionVal.None
+    var lessRecent, moreRecent: OptionVal[Node[A]] = OptionVal.None
+  }
+}
+
+/**
+ * INTERNAL API
+ *
+ * Mutable non-thread-safe frequency list.
+ * Used for tracking frequency of elements for implementing least/most frequently used eviction policies.
+ * Implemented using a doubly-linked list of doubly-linked lists, with lookup, so that all operations are constant time.
+ */
+@InternalApi
+private[akka] final class FrequencyList[A] {
+  import FrequencyList.{ FrequencyNode, Node }
+
+  private var leastFrequent, mostFrequent: OptionVal[FrequencyNode[A]] = OptionVal.None
+  private val lookupNode = mutable.Map.empty[A, Node[A]]
+
+  def size: Int = lookupNode.size
+
+  def update(value: A): FrequencyList[A] = {
+    if (lookupNode.contains(value)) {
+      val node = lookupNode(value)
+      increaseFrequency(node)
+    } else {
+      val node = new Node(value)
+      addAsLeastFrequent(node)
+      lookupNode += value -> node
+    }
+    this
+  }
+
+  def remove(value: A): FrequencyList[A] = {
+    if (lookupNode.contains(value)) {
+      val node = lookupNode(value)
+      removeFromCurrentPosition(node)
+      lookupNode -= value
+    }
+    this
+  }
+
+  def contains(value: A): Boolean = lookupNode.contains(value)
+
+  private def leastNode: OptionVal[Node[A]] =
+    if (leastFrequent.isDefined) leastFrequent.get.leastRecent else OptionVal.None
```

Comment:
you can use `match` without any allocations
```
leastFrequent match {
  case OptionVal.Some(node) => node.leastRecent
  case _ => OptionVal.None
```

Reply:
cool, I've updated to use `match` instead in a few places

Reply:
Actually, I've now switched over to a circular linked list approach, with a "zero" node. Makes the linking/unlinking and other logic simpler, without all the OptionVal / null checks. https://github.com/akka/akka/pull/30898/commits/2b723bb5584d1323c1a39f0bc2b5605eff1a4137